modules/calib3d/src/calibinit.cpp

```diff
@@ -2266,7 +2266,7 @@ bool findCirclesGrid( InputArray _image, Size patternSize,
         }
     }
 
-    if (!H.empty())  // undone rectification
+    if (isValid && !H.empty())  // undone rectification
```

Comment:
relates #19079

The problem is that [original code](https://github.com/opencv/opencv/blob/3.4.11/modules/calib3d/src/calibinit.cpp#L2283-L2284) returns [transformed](https://github.com/opencv/opencv/blob/3.4.11/modules/calib3d/src/calibinit.cpp#L2280) (rectified) corners for the "false" result too.
Why do we need to return these rectified corners on detection failure?

Please provide more details about observed issue. What error message did you get?

BTW, you can see this corners problem by launching calib3d test with and without your patch:
```
./bin/opencv_test_calib3d --gtest_filter=*regression_18713* --test_debug=10
```

Reply:
Hi @alalek, Yep #19079 introduced the issue.

Since the rectification code was removed from the `isFound` conditional in the main attempts for loop, we need to add back in that check before attempting to rectify the centers. 

The error occurs when the attempts loop reaches its last attempt. If no centers are found on the last attempt, but centers were found on one of the previous attempts, `H` will be valid but `centers` will be empty. The final `!H.empty()` conditional will then attempt to transform a empty container of centers which will trigger the following assertion:
```
opencv/modules/core/src/matmul.dispatch.cpp:439: error: (-215:Assertion failed) scn == m.cols || scn + 1 == m.cols in function 'transform'
```

I attempted to preserve the original function interface, and figured checking if the result was valid before rectifying the centers was the best way to do that.

Reply:
Perhaps I got this. Probably we should check for `corners.empty()` instead.

Please capture `points` values (`std::cout << Mat(points) << std::endl`) which cause this error, so we can add regression test for that.
(add `patternSize` / `flags` arguments values too)

Reply:
![debug](https://user-images.githubusercontent.com/1621924/107726000-9c2ce380-6cb5-11eb-8005-560094d3364e.png)

I figured `isValid` was equivalent to `centers.empty()`, but I am happy to change this to check the centers directly. I attached the image which failed for reference. I believe the partial circle grid is what throws the logic off.

I am new to the unit testing framework for OpenCV. What is the best way to add the test for detectors like this? Should I hardcode the image contents in the test?

The `patternSize` is used was 13x8 and I used the default args for `flags` and `blobDetector`. Those are `CALIB_CB_SYMMETRIC_GRID` and `SimpleBlobDetector::create()` respectively

Reply:
I just took a look at the `regression_18713` test. I port this issue to a similar test